test/micro/org/openjdk/bench/java/net/URLEncodeDecode.java

```diff
@@ -51,78 +52,140 @@
+            if (s.equals(java.net.URLEncoder.encode(s, StandardCharsets.UTF_8))) {
+                countUnchanged++;
+            } else {
+                if (unchanged == 100) {
+                    System.out.println("Unexpectedly needs encoding action: ");
+                    System.out.println("\t" + s);
+                    System.out.println("\t" + java.net.URLEncoder.encode(s, StandardCharsets.UTF_8));
+                }
+            }
+        }
+        System.out.println();
+        System.out.println("Generated " + testStringsEncode.length + " encodable strings, " + countUnchanged + " of which does not need encoding action");
 
-        for (int i = 0; i < count; i++) {
+        for (int i = 0; i < COUNT; i++) {
             int l = r.nextInt(maxLength);
+            boolean needDecoding = r.nextInt(100) >= unchanged;
             StringBuilder sb = new StringBuilder();
+            boolean hasDecoded = false;
             for (int j = 0; j < l; j++) {
-                int c = r.nextInt(tokens.length + 5);
-                if (c >= tokens.length) {
-                    sb.append("%").append(tokens[r.nextInt(16)]).append(tokens[r.nextInt(16)]);
+                if (needDecoding && r.nextInt(100) < encodeChars) {
+                    addDecodableChar(tokens, r, sb);
+                    hasDecoded = true;
                 } else {
-                    sb.append(tokens[c]);
+                    addToken(tokens, r, sb);
                 }
             }
+            if (needDecoding && !hasDecoded) {
+                addDecodableChar(tokens, r, sb);
+            }
             testStringsDecode[i] = sb.toString();
         }
+        countUnchanged = 0;
+        for (String s : testStringsDecode) {
+            if (s.equals(java.net.URLDecoder.decode(s, StandardCharsets.UTF_8))) {
+                countUnchanged++;
+            } else {
+                if (unchanged == 100) {
+                    System.out.println("Unexpectedly needs encoding action: ");
+                    System.out.println("\t" + s);
+                    System.out.println("\t" + java.net.URLDecoder.decode(s, StandardCharsets.UTF_8));
+                }
+            }
+        }
+        System.out.println("Generated " + testStringsDecode.length + " decodable strings, " + countUnchanged + " of which does not need decoding action");
+    }
+
+    private static void addToken(char[] tokens, Random r, StringBuilder sb) {
+        int c = r.nextInt(tokens.length);
+        sb.append(tokens[c]);
+    }
+
+    private static void addDecodableChar(char[] tokens, Random r, StringBuilder sb) {
+        if (r.nextInt(100) < 15) {
+            sb.append('+'); // exercise '+' -> ' ' decoding paths.
+        } else {
+            sb.append("%").append(tokens[r.nextInt(16)]).append(tokens[r.nextInt(16)]);
+        }
     }
 
     @Benchmark
     public void testEncodeUTF8(Blackhole bh) throws UnsupportedEncodingException {
         for (String s : testStringsEncode) {
```

Comment:
The consume is pretty cheap with compiler assisted blackholes, but can still overall be too noisy with small 's'

Worth running perfasm and see what it says and/or reading put of the encoded value and extract some property to accumulate in the for loop. And see if is less costy.

Another point instead is related loop unrolling/strip mining; safepoint polls can create assitionalrnandom noise depending how much loop unrolling is excercized (Vs the cost of encoding): it worth to enforce a specific unrolling level or disabling it as a whole, too, to get more stable results.

Reply:
I think these are all valid concerns and it's a worthwhile endeavor to discuss and try and improve how we benchmark loops (or more specifically variable inputs). However I think this is out of scope for this improvement - which is mainly about correctness of data and making sure we're testing the right thing - and ought to be discussed and experimented with separately. If you don't strongly disagree I will integrate this so we can move ahead with other work in this area (#15354 for starters)